python/generate_compile_commands.py

```diff
@@ -0,0 +1,102 @@
+import os
+import json
+
+
+def get_filelist(root, whitelist):
+    """
+    1. Generates a file list dictionary of the structure:
+
+    {
+        pcl_folder: {
+            sub_folder: {
+                "impls": [list_of_impl_files]
+                "headers": [list_of_header_files]
+            }
+            ...
+        }
+        ...
+    }
+
+    """
+    pcl = {}
+
+    for folder in whitelist:
+        folder_path = os.path.join(root, folder)
+        if os.path.exists(folder_path):
+            impl_files = []
+            header_files = []
+
+            # walk the folder to get filenames
+            for root, _, files in os.walk(folder_path):
+                for file in files:
+                    # filter for implementation files
+                    if file.split(".")[-1] == "hpp":
+                        impl_files.append(os.path.join(root, file))
+                    # filter for header files
+                    if file.split(".")[-1] == "h":
+                        header_files.append(os.path.join(root, file))
+                    # add a filter for any other file type if desired
+                    _
+
+            # add list to pcl dict
+            pcl[folder] = {"impls": impl_files, "headers": header_files}
+
+        else:
+            raise Exception(f"{folder_path} doesn't exist")
+
+    return pcl
+
+
+def generate_compile_commmands(
```

Comment:
Please document the variables and their types since it's not apparent for file_list and compiler arguments (rest all are clear-ish, but there's pathlib.Path or string for path also)

Reply:
https://docs.python.org/3/library/typing.html